packages/react/src/SelectPanel/SelectPanel.tsx

```diff
@@ -87,19 +93,69 @@ export function SelectPanel({
   textInputProps,
   overlayProps,
   sx,
+  loading,
+  initialLoadingType = 'spinner',
   ...listProps
 }: SelectPanelProps): JSX.Element {
+  const inputRef = React.useRef<HTMLInputElement>(null)
   const titleId = useId()
   const subtitleId = useId()
+  const dataLoadedOnce = useRef(false)
+  const [isLoading, setIsLoading] = useProvidedStateOrCreate(loading, undefined, false)
   const [filterValue, setInternalFilterValue] = useProvidedStateOrCreate(externalFilterValue, undefined, '')
+  const {safeSetTimeout, safeClearTimeout} = useSafeTimeout()
+  const loadingDelayTimeoutId = useRef<number | null>(null)
   const onFilterChange: FilteredActionListProps['onFilterChange'] = useCallback(
     (value, e) => {
+      if (dataLoadedOnce.current) {
+        // If data has already been loaded once, delay the spinner a bit. This also helps
+        // not show and then immediately hide the spinner if items are loaded quickly, i.e.
+        // not async.
+
+        if (loadingDelayTimeoutId.current) {
+          safeClearTimeout(loadingDelayTimeoutId.current)
+        }
+
+        loadingDelayTimeoutId.current = safeSetTimeout(() => setIsLoading(true), 1000)
+      } else {
+        // If this is the first data load and there are no items, show the loading spinner
+        // immediately
+
+        if (items.length === 0) {
+          setIsLoading(true)
+        }
+      }
+
       externalOnFilterChange(value, e)
       setInternalFilterValue(value)
     },
-    [externalOnFilterChange, setInternalFilterValue],
+    [externalOnFilterChange, setInternalFilterValue, setIsLoading, safeSetTimeout, safeClearTimeout, items],
   )
 
+  useEffect(() => {
+    if (isLoading) {
+      setIsLoading(false)
+      dataLoadedOnce.current = true
+    }
+    // Only fire this effect if items have changed
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [items])
+
+  // Populate panel with items on first open
+  useEffect(() => {
+    // If data was already loaded once, do nothing
+    if (dataLoadedOnce.current) return
+
+    // Only load data when the panel is open
+    if (open) {
+      // Only trigger filter change event if there are no items
+      if (items.length === 0) {
+        // Trigger filter event to populate panel on first open
+        onFilterChange(filterValue, null)
+      }
+    }
+  }, [open, dataLoadedOnce, onFilterChange, filterValue, items])
```

Comment:
Can you explain a bit what we are doing here, I wasn't able to understand

Thinking out loud: when there is no data yet (`dataLoadedOnce.current = false` and `items.length = 0`)? why though?

Reply:
Yes! We do that to trigger loading items when the panel first opens. The way to trigger loading items is to tell the consumer that the filter has changed. If there are already items provided, then don't trigger loading items because they're presumably already loaded.